Update Tracing API: IsRecording wording (#3020)
open-telemetry/opentelemetry-specification: specification/trace/api.md

#### IsRecording

A `Span` is recording (`IsRecording` returns `true`) when the data provided to
it via functions like `SetAttributes`, `AddEvent`, `SetStatus` is captured in
some form (e.g. in memory). When a `Span` is not recording (`IsRecording` returns
`false`), all this data is discarded right away. Further attempts to set or add
data will not record, making the span effectively a no-op.

> This flag may be `true` despite the entire trace being sampled out. This

Reviewer: ```suggestion
This flag may be `true` despite the entire trace not being sampled. This
```

> allows to record and process information about the individual Span without

Reviewer 2: ```suggestion
allows information about the individual Span to be recorded and processed without
```

sending it to the backend. An example of this scenario may be recording and
processing of all incoming requests for the processing and building of
SLA/SLO latency charts while sending only a subset - sampled spans - to the
backend. See also the [sampling section of SDK design](sdk.md#sampling).

> After a `Span` is ended, it SHOULD become non-recording and its `IsRecording`
> SHOULD start consequently return `false`. The one known exception to this is

Reviewer 2: ```suggestion
After a `Span` is ended, it SHOULD become non-recording and `IsRecording`
SHOULD always return `false`. The one known exception to this is
```

> SHOULD start consequently return `false`. The one known exception to this is

Reviewer 3: ```suggestion
SHOULD consequently return `false`. The one known exception to this are
```

Author: This is & are twisting me. Are you sure it is "are"?

Reviewer 2: "The one known exception" is the singular subject, "is" correct in this case.

#### IsRecording

> Returns `true` if this `Span` has been started but not ended.

Reviewer 2: This statement does not take into account spans that are sampled to be non-recording. A span being started is not the only criteria for a span to be recorded.

Author: Maybe I misinterpret your first sentence but I think sampled but non-recording is against the spec. https://opentelemetry.io/docs/reference/specification/trace/sdk/#shouldsample `The flag combination SampledFlag == true and IsRecording == false could cause gaps in the distributed trace, and because of this OpenTelemetry API MUST NOT allow this combination.`

My patch attempt is by me reverse engineering the golang SDK. Maybe the golang implementation is not complete but otherwise these 2 variants are how the API is implemented.

```
// IsRecording returns if this span is being recorded. If this span has ended
// this will return false.
func (s *recordingSpan) IsRecording() bool {
	if s == nil {
		return false
	}
	s.mu.Lock()
	defer s.mu.Unlock()

	return s.endTime.IsZero()
}
```

And if it is a non-recording span:

```
// IsRecording always returns false.
func (nonRecordingSpan) IsRecording() bool { return false }
```

Maybe something like this accounts the non-recording spans.
`Returns true for recording spans if this Span has been started but not ended. Returns false after the Span is ended or if the Span is a non recording span.`

Let me know what you think please.

Reviewer 3: If you start a span but the sampler returns a negative sampling decision, the SDK will return a non-recording span that is not recording and not sampled. You still have a span in a "started" state, so the wording proposed here saying just "started but not ended" will not work, as it does not account for such non-recording spans (the second code snippet you cited).

See also https://opentelemetry.io/docs/reference/specification/trace/sdk/#recording-sampled-reaction-table and https://opentelemetry.io/docs/reference/specification/trace/sdk/#shouldsample ("Return value").

Author: I believe I have updated the patch to reflect your comment. I believe it is you who will mark the conversation as "resolved".

Reviewer 3: Thanks, Umut!

While this is correct now, it references implementation details from the SDK specification and actually duplicates these statements, making the spec harder to maintain and breaking the distinction made on purpose. The API spec you are editing here defines what IsRecording means to users of the API, while the SDK spec defines what SDKs should do and return based on the sampling decision.
It might be a bit confusing at first but I think we should not mix these two together here.

If there is any clarification that can be made without referencing the SDK but defining better how to _interpret_ the values returned for IsRecording, feel free to propose something.

Author: Thanks again. I just removed the the reference to the SDK. @yurishkuro has also suggested the same thing.

#### IsRecording

> Always returns `false` If this is a non-recording `Span` (Sampling

Author: I am blank regarding how to address your comment. Any suggestions? 

We do mention about the behavior when recording and sampling flags are set under the section Sampling in the SDK. Maybe the SDK should be made more clear to reflect memory Vs export.

> Always returns `false` If this is a non-recording `Span` (Sampling

Reviewer 4: IsRecording = false should mean that the data is discarded right away, i.e. nobody will (be able to) read it, not even in memory. It's an effectively no-op span.

`Decision` was `DROP` at the time of `Span` creation). Otherwise returns
`true` if this `Span` has been started but not ended.
After a `Span` is ended, it usually becomes non-recording and thus

> `IsRecording` SHOULD consequently return `false` for ended Spans.

Reviewer 5: (side remark: The appearance of "usually" is ... unusual without a corresponding SHOULD statement. I wonder if we should replace "usually" with "SHOULD" and if not, why not?)

Author: English is not my first language but shouldn't SHOULDs be SHALLS? And why upper case? 

I have been quite confused when I read this paragraph. What I got confused is "`it usually becomes non-recording...`" So there is an exception, seems like streaming applications. Then we have this: "`are one expected case where IsRecording cannot change after ending a Span.`" I have not seen this case in go SDK implementation. 

With my limited knowledge, I can agree with what you propose. "`After a Span is ended, it SHOULD become non-recording..`"

Reviewer 3: @utezduyar You can find details on the distinction and the casing in https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification#notation-conventions-and-compliance and the RFCs it refers to.

Note: Streaming implementations, where it is not known if a span is ended,
are one expected case where `IsRecording` cannot change after ending a Span.

Recording information like events with the `AddEvent` operation, attributes

> using `SetAttributes`, status with `SetStatus`, etc will not perform if

Reviewer 5: ```suggestion
using `SetAttributes`, status with `SetStatus`, etc will not perform when
```
